tests/testthat/test-forge-formula.R

```diff
@@ -346,17 +495,27 @@ test_that("novel levels are ignored correctly when the new column is a character
     stringsAsFactors = FALSE
   )
 
-  blueprint <- default_formula_blueprint(allow_novel_levels = TRUE)
+  bp1 <- default_formula_blueprint(allow_novel_levels = TRUE)
+  bp2 <- default_formula_blueprint(allow_novel_levels = TRUE, composition = "dgCMatrix")
 
-  x <- mold(y ~ f, dat, blueprint = blueprint)
+  x1 <- mold(y ~ f, dat, blueprint = bp1)
+  x2 <- mold(y ~ f, dat, blueprint = bp2)
 
   expect_warning(
-    xx <- forge(new, x$blueprint),
+    xx1 <- forge(new, x1$blueprint),
+    NA
+  )
+  expect_warning(
+    xx2 <- forge(new, x2$blueprint),
     NA
   )
 
   expect_equal(
-    xx$predictors[[5,5]],
+    xx1$predictors[[5,5]],
+    1
+  )
+  expect_equal(
+    unname(xx2$predictors[5,5]),
```

Comment:
Does `[[` do something different with sparse matrices?

Reply:
Yes, they do not subset at that level:

``` r
library(Matrix)
n <- 1e5 
y <- sample(0:1, n, replace = TRUE) 
x1 <- rnorm(n)
x2 <- sample(1:400, n, replace = TRUE)
X1 <- sparse.model.matrix(~ x1 + factor(x2))
X1[[5,5]]
#> Error in X1[[5, 5]]: incorrect number of subscripts
X1[5,5]
#> [1] 0
```

<sup>Created on 2020-10-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>